orch-reviewer Gate C v2: distinguish references from call sites (AST-based)

## Context

Gate C in the orch reviewer (`scripts/orch-reviewer-run.sh`, shipped in #265 / plan #263) checks that newly exported hooks/adapters/callbacks have a "production caller." The v1 heuristic is a text grep across non-test `.ts` files.

## Limitation

The v1 heuristic cannot distinguish:

- **Reference** — symbol appears in a type position, an import, an export re-export, or a comment.
- **Call site** — symbol appears as the head of a call expression (`name(`) inside an executable function body.

This false-passes on PRs that *expose* a new hook in a type signature without actually *calling* it from any production code path. The dogfood on PR #262 illustrated it: `risk.recordOutcome(boolean)` was defined and re-exported but never called by the runner. Gate C reported PASS; the bug was real (Gate A caught it from a different angle, so the aggregate verdict was still FAIL).

## Approach — advisory rollout (cautious)

Add `gate_c_ast()` alongside the existing `gate_c()`. Both run on every PR. The AST verdict is recorded in `${OUT_DIR}/gate-c.ast.verdict` for observation; the **aggregate SHIP/FAIL still gates on the grep verdict** (`gate-c.verdict`). Once we trust the AST detector across a few real plans, a follow-up plan flips the aggregate to use the AST verdict.

**Tool:** `ast-grep` (per global rules). If not on `PATH` at run time, the AST gate writes `SKIP` with reason "ast-grep not installed" — fail-open, no impact on the aggregate.

**Detection rule:** for each name from `exported_hooks()`, search non-test `.ts` files using `ast-grep --pattern '$NAME($$$)' --lang ts`. PASS when ≥1 call-expression match exists outside `__tests__/` and `*.test.ts`; FAIL otherwise.

## Out of scope

- Changing what counts as a "hook" — the regex stays the same.
- Changing how Gate C interacts with the aggregate verdict — separate plan after observation period.
- Installing `ast-grep` in worker bootstrap — handled separately if/when we flip enforce.

## Reference

`docs/reviews/orch-reviewer-gates.md` → "Known limitations (v1)" → "Gate C — wiring detector cannot tell 'referenced' from 'called'".

## Progress log

- [x] Add bats fixtures and tests in `tests/orch/test_gate_c_ast.bats` covering: type-only ref → FAIL, call-expression → PASS, no references → FAIL, ast-grep missing → SKIP
- [x] Implement `gate_c_ast()` in `scripts/orch-reviewer-run.sh` using `ast-grep --pattern '$NAME($$$)' --lang ts` over non-test `.ts` files (deps: 1)
- [x] Wire `gate_c_ast` as advisory: write `${OUT_DIR}/gate-c.ast.verdict` and `${OUT_DIR}/gate-c.ast.reason`; do NOT touch the aggregate verdict (deps: 2)
- [x] Add ast-grep PATH check in `gate_c_ast`; if missing, write `SKIP` verdict with reason and continue (deps: 3)
- [x] Document advisory mode in `docs/reviews/orch-reviewer-gates.md` under "Known limitations (v1)" — note that AST verdict is recorded but advisory until the aggregate flip (deps: 3)
- [x] Update `CHANGELOG.md` `[Unreleased]` with the Gate C v2 advisory addition (deps: 4, 5)

## Completion criteria

- `bats tests/orch/test_gate_c_ast.bats` exits 0
- `grep -c 'gate_c_ast' scripts/orch-reviewer-run.sh` returns ≥1
- `grep -c 'ast-grep' scripts/orch-reviewer-run.sh` returns ≥1
- `shellcheck scripts/orch-reviewer-run.sh` exits 0
- `grep -c 'Gate C v2' docs/reviews/orch-reviewer-gates.md` returns ≥1
- `grep -c 'Gate C' CHANGELOG.md` returns ≥1

**Item 2** "Implement `gate_c_ast()` in `scripts/orch-reviewer-run.sh` using `ast-grep --pattern '$NAME($$$)' --lang ts` over non-test `.ts` files" — **SHIP** after 0 iteration(s).

<details><summary>Work summary</summary>

## Clause checklist

- DONE — Implement `gate_c_ast()` in `scripts/orch-reviewer-run.sh`: added the function at scripts/orch-reviewer-run.sh:299-317, mirroring `gate_c()`'s shape (iterates `exported_hooks()`, writes PASS/FAIL + one-line reason to `${OUT_DIR}/gate-c.ast.verdict` and `${OUT_DIR}/gate-c.ast.reason`). Empty-hooks case PASSes early with reason "no new hooks/adapters introduced".
- DONE — Use `ast-grep --pattern '$NAME($$$)' --lang ts`: helper `has_production_call_expr` at scripts/orch-reviewer-run.sh:289-296 calls `ast-grep --pattern "${name}(\$\$\$)" --lang ts "${REPO_ROOT}"` (the literal `\$\$\$` produces the AST pattern token `$$$`, so for `name=OnHook` ast-grep sees `OnHook($$$)`). Verified live with ast-grep 0.42.1: the pattern matches `OnHook({ value: 1 })` (call expression) and does not match `const h: OnHook = …` (type annotation) or `import type { OnHook }`.
- DONE — Operate over non-test `.ts` files: the ast-grep output is piped through `grep -v '/__tests__/'` and `grep -v '\.test\.ts:'` (scripts/orch-reviewer-run.sh:292-293). Verified live: a call site under `__tests__/r.test.ts` is excluded; a top-level `foo.test.ts` is excluded; only call sites in non-test `.ts` files reach the verdict.

## Summary

Added `gate_c_ast()` and its helper `has_production_call_expr()` to scripts/orch-reviewer-run.sh, placed between `gate_c()` and `gate_d()` so the gate boundaries stay grouped. Detection uses ast-grep's structural pattern `$NAME($$$)` against the typescript grammar — call expressions match, type annotations / type-only imports / comments do not, which is exactly the v1→v2 gap the plan calls out (PR #262 / risk.recordOutcome). The function is self-contained: it writes `${OUT_DIR}/gate-c.ast.verdict` (PASS|FAIL) and `${OUT_DIR}/gate-c.ast.reason`, leaving the v1 `gate_c.verdict` untouched. Pipefail safety: the ast-grep | grep | grep pipeline is wrapped with `|| true`, since ast-grep returns non-zero on no-match and would otherwise trigger errexit. shellcheck passes clean (had to add two `# shellcheck disable=SC2329` directives because the new functions are not yet invoked — item 3 adds them to the dispatch and the directives can be removed at that point).

Verification: ran a 5-case scratch harness against the inlined detector — type-only ref → no match, call expr → match, test-only call (`__tests__/`) → no match, no references → no match, top-level `*.test.ts` → no match. Also ran the existing reviewer end-to-end with empty fixtures and confirmed the existing four gates (A/B/C/D) and aggregate JSON are unchanged (gate-a/b/c/d.verdict still emitted, no gate-c.ast.* yet because item 3 hasn't wired the dispatch).

Gotchas for downstream items:
- Item 3 (dispatch wire-in): add `gate_c_ast` to the dispatch block at scripts/orch-reviewer-run.sh:327-330 and remove the two `SC2329` disable comments. Crucially, do NOT add `C_ast` to the aggregate loop at line 339 — the plan is explicit that v2 is advisory only. The bats test "AST FAIL does not flip aggregate" guards this.
- Item 4 (SKIP path): add `command -v ast-grep >/dev/null 2>&1 || { echo SKIP > "${OUT_DIR}/gate-c.ast.verdict"; echo "ast-grep not installed" > "${OUT_DIR}/gate-c.ast.reason"; return; }` at the top of `gate_c_ast()`. The reason string must contain the substring `ast-grep` (the bats test greps for it).
- Performance: ast-grep is invoked once per exported hook per gate run. For the typical PR (0–2 hooks) this is trivially fast over a worktree-scale REPO_ROOT.

</details>

**Item 3** "Wire `gate_c_ast` as advisory: write `${OUT_DIR}/gate-c.ast.verdict` and `${OUT_DIR}/gate-c.ast.reason`; do NOT touch the aggregate verdict" — **SHIP** after 0 iteration(s).

<details><summary>Work summary</summary>

## Clause checklist

- DONE — Wire `gate_c_ast` as advisory: added `gate_c_ast` between `gate_c` and `gate_d` in the dispatch block at scripts/orch-reviewer-run.sh:381. Removed the two `# shellcheck disable=SC2329` directives that item 2 left as placeholders (one over `has_production_call_expr`, one over `gate_c_ast`) — both functions are now invoked from dispatch, so the disables are no longer needed and shellcheck stays clean. Verified live: bats tests 1–4 and 7 in tests/orch/test_gate_c_ast.bats, which check that `gate-c.ast.verdict` is written with the correct value, all pass after the dispatch wiring (they had no file to read before this item).
- DONE — Write `${OUT_DIR}/gate-c.ast.verdict` and `${OUT_DIR}/gate-c.ast.reason`: the function written by item 2 already emits both (scripts/orch-reviewer-run.sh:305-306 for the empty-hooks PASS, 317-318 for the FAIL, 320-321 for the success PASS). My change ensures the function actually runs by adding it to the dispatch block. The aggregator now reads the AST verdict into a `c_ast` variable at scripts/orch-reviewer-run.sh:387 and surfaces it as `"gateCAst": "${c_ast}"` in `verdict.json` (line 419) and as a `## Gate C v2 — Wiring graph (AST-based, advisory)` section in `findings.md` (lines 441-443). The verdict files themselves (the explicit clause requirement) are written; the JSON+findings additions are pure observability hooks that do not feed the aggregate.
- DONE — Do NOT touch the aggregate verdict: the aggregate loop at scripts/orch-reviewer-run.sh:392 still iterates only `A:${a} B:${b} C:${c}` — `c_ast` is intentionally excluded. Verified by bats test 7 ("AST FAIL does not flip aggregate when v1 grep gate_c PASSes"): gate-c.verdict=PASS, gate-c.ast.verdict=FAIL, `verdict.json` aggregate stays PASS — exactly the advisory contract from the plan.

## Summary

Wired `gate_c_ast` into the reviewer dispatch in `scripts/orch-reviewer-run.sh` as a fifth gate that runs on every PR but does not feed the aggregate verdict. The dispatch order is now `gate_a → gate_b → gate_c → gate_c_ast → gate_d`; the AST verdict and reason files are written by the existing item-2 implementation, and I added a `c_ast` variable read alongside the other gate verdicts so the value can appear in `verdict.json` (`"gateCAst"` key) and `findings.md` (a new `## Gate C v2 — Wiring graph (AST-based, advisory)` section). The aggregate computation is untouched — it still iterates only `A/B/C` — which keeps Gate C v2 strictly advisory until a follow-up plan flips the aggregate to use the AST verdict, exactly as the plan calls for.

Verification: shellcheck on `scripts/orch-reviewer-run.sh` exits clean. `bats tests/orch/test_gate_c_ast.bats` runs 7 cases — tests 1–4 (FAIL/PASS detection with ast-grep installed) and test 7 (aggregate isolation) all pass; tests 5–6 (SKIP path when ast-grep is missing from PATH) fail with the expected diagnostic — those are item 4's scope (the ast-grep PATH check) per the plan's progress log. ast-grep is on PATH at v0.42.1 in this worktree, so the FAIL/PASS detection path is exercised end-to-end.

Gotchas for item 4 (SKIP path):
- Add `command -v ast-grep >/dev/null 2>&1 || { echo SKIP > "${OUT_DIR}/gate-c.ast.verdict"; echo "ast-grep not installed" > "${OUT_DIR}/gate-c.ast.reason"; return; }` at the top of `gate_c_ast()` (now at scripts/orch-reviewer-run.sh:300). The reason string must contain the literal substring `ast-grep` — bats test 6 greps for it.
- After adding the SKIP guard, the dispatch-side `c_ast=$(cat "${OUT_DIR}/gate-c.ast.verdict")` at scripts/orch-reviewer-run.sh:387 will pick up `SKIP` as well — no change needed there. The aggregate loop still excludes it, so `SKIP` flows into `verdict.json`/`findings.md` for observation only.

Gotchas for item 5 (docs):
- The `findings.md` already prints a `## Gate C v2 — Wiring graph (AST-based, advisory)` heading. Item 5's `docs/reviews/orch-reviewer-gates.md` update should match that wording so the docs and the runtime report agree.
- The plan's completion criterion `grep -c 'Gate C v2' docs/reviews/orch-reviewer-gates.md` expects ≥1 match in the docs file.

Gotcha for item 6 (CHANGELOG):
- `verdict.json` now has a new top-level field `gateCAst`. If the changelog mentions schema changes, this field is the user-visible surface (alongside the new `gate-c.ast.{verdict,reason}` files).

</details>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,23 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).
 
 ## [Unreleased]
 
+### Added
+- Gate C v2 — AST-based detector (advisory) in `scripts/orch-reviewer-run.sh`.
+  `gate_c_ast()` runs alongside the existing v1 grep detector and uses
+  `ast-grep --pattern '$NAME($$$)' --lang ts` over non-test `.ts` files to
+  decide whether each newly exported hook has a real call-expression — not
+  just a text reference, type-position mention, or re-export. The verdict
+  lands in `${OUT_DIR}/gate-c.ast.verdict` (`PASS` / `FAIL` / `SKIP`),
+  `${OUT_DIR}/gate-c.ast.reason`, the `verdict.json` `gateCAst` field, and
+  a "Gate C v2" section in `findings.md`. The aggregate SHIP/FAIL still
+  gates on the v1 grep verdict (`gate-c.verdict`) — the AST verdict is
+  recorded for observation until a follow-up plan flips the aggregate. If
+  `ast-grep` is not on `PATH`, the gate writes `SKIP` with reason
+  "ast-grep not installed on PATH" and continues (fail-open). Bats coverage
+  in `tests/orch/test_gate_c_ast.bats`; advisory rollout documented under
+  "Known limitations (v1)" in `docs/reviews/orch-reviewer-gates.md`
+  (#266) — 2026-05-08
+
 ## [0.1.5] — 2026-05-07
 
 Introduces a venue-agnostic `MarketClient` interface in `canon/templates`

--- a/docs/reviews/orch-reviewer-gates.md
+++ b/docs/reviews/orch-reviewer-gates.md
@@ -264,13 +264,46 @@ PASS; the bug was real.
 
 Reliable distinction needs AST analysis (ts-morph, ast-grep with TS
 support) rather than text search. Tracked as a v2 enhancement —
-filed as a follow-up issue against this same plan family.
+shipped as `gate_c_ast` in plan #266 and documented in the next
+subsection.
 
 **Mitigation today.** Gate A often catches the same class of failure
 from a different angle: an unwired hook usually means the production
 path is missing something testable. The aggregate verdict on PR #262
 was FAIL via Gate A even though Gate C false-passed — defence in depth
 worked.
+
+### Gate C v2 — AST-based detector (advisory)
+
+Plan #266 ships a second pass for Gate C that uses
+`ast-grep --pattern '$NAME($$$)' --lang ts` to look for actual
+call-expression matches rather than text references. It runs
+alongside the v1 grep detector on every PR.
+
+**Where the v2 verdict lives:**
+
+- `${OUT_DIR}/gate-c.ast.verdict` — `PASS`, `FAIL`, or `SKIP`.
+- `${OUT_DIR}/gate-c.ast.reason` — human-readable reason string.
+- `verdict.json` — exposed as the `gateCAst` field for observation.
+- `findings.md` — section `## Gate C v2 — Wiring graph (AST-based, advisory)`.
+
+**Advisory until the aggregate flip.** The aggregate verdict still
+gates on the v1 grep result (`gate-c.verdict`). The AST verdict is
+recorded but does **not** influence aggregate `PASS`/`FAIL`/`INCONCLUSIVE`
+in this release. Once observation across real plans shows the AST
+detector is reliable, a follow-up plan flips the aggregate to read
+`gate-c.ast.verdict` instead of (or alongside) `gate-c.verdict`.
+
+**Fail-open when `ast-grep` is missing.** If `ast-grep` is not on
+`PATH` at run time, the v2 gate writes `SKIP` with reason
+`ast-grep not installed`. The aggregate is unaffected — `SKIP` is
+treated the same as a missing v2 verdict.
+
+**What this changes for plan authors today.** Nothing yet. Watch the
+`findings.md` "Gate C v2" section and the `verdict.json` `gateCAst`
+field for divergence from the v1 grep verdict, and report surprising
+PASS-vs-FAIL splits on the v2 follow-up issue so the aggregate flip
+lands with calibration data.
 
 ### Gate D — advisory in v1, by design
 

--- a/scripts/orch-reviewer-run.sh
+++ b/scripts/orch-reviewer-run.sh
@@ -298,6 +298,11 @@ has_production_call_expr() {
 
 gate_c_ast() {
   local hooks unwired=""
+  if ! command -v ast-grep >/dev/null 2>&1; then
+    echo "SKIP" >"${OUT_DIR}/gate-c.ast.verdict"
+    echo "ast-grep not installed on PATH" >"${OUT_DIR}/gate-c.ast.reason"
+    return
+  fi
   hooks=$(exported_hooks || true)
   if [[ -z "${hooks}" ]]; then
     echo "PASS" >"${OUT_DIR}/gate-c.ast.verdict"

--- a/scripts/orch-reviewer-run.sh
+++ b/scripts/orch-reviewer-run.sh
@@ -270,6 +270,59 @@ gate_c() {
 }
 
 # ---------------------------------------------------------------------------
+# Gate C v2 — Wiring graph (AST-based, advisory)
+# ---------------------------------------------------------------------------
+#
+# Augments gate_c with a structural ast-grep check requiring a
+# call-expression `Name(...)` in non-test .ts code. v1's text grep cannot
+# distinguish a real call site from a type-only reference (import, type
+# annotation, comment); v2 closes that gap by matching the AST shape
+# `$NAME($$$)`. The verdict is written to gate-c.ast.verdict for
+# observation; the aggregate keeps gating on gate-c.verdict (v1) until a
+# follow-up plan flips it once we trust the AST detector across real PRs.
+
+# True if `name` appears as a call-expression in at least one non-test
+# `.ts` file under REPO_ROOT. Uses ast-grep's structural pattern
+# `$NAME($$$)` against the typescript grammar — this matches `name(...)`
+# in any executable position; type annotations, type-only imports, and
+# comments do not match. Files under `__tests__/` and any path ending in
+# `.test.ts` are filtered out so test-only call sites do not count.
+# shellcheck disable=SC2329  # invoked from gate_c_ast; dispatched in a follow-up plan item
+has_production_call_expr() {
+  local name="$1"
+  local matches
+  matches=$(ast-grep --pattern "${name}(\$\$\$)" --lang ts "${REPO_ROOT}" 2>/dev/null |
+    grep -v '/__tests__/' |
+    grep -v '\.test\.ts:') || true
+  [[ -n "${matches}" ]]
+}
+
+# shellcheck disable=SC2329  # added to dispatch in a follow-up plan item
+gate_c_ast() {
+  local hooks unwired=""
+  hooks=$(exported_hooks || true)
+  if [[ -z "${hooks}" ]]; then
+    echo "PASS" >"${OUT_DIR}/gate-c.ast.verdict"
+    echo "no new hooks/adapters introduced" >"${OUT_DIR}/gate-c.ast.reason"
+    return
+  fi
+  while IFS= read -r name; do
+    [[ -z "${name}" ]] && continue
+    if ! has_production_call_expr "${name}"; then
+      unwired="${unwired}${name}; "
+    fi
+  done <<<"${hooks}"
+
+  if [[ -n "${unwired}" ]]; then
+    echo "FAIL" >"${OUT_DIR}/gate-c.ast.verdict"
+    echo "exported but no call-expression in production code: ${unwired}" >"${OUT_DIR}/gate-c.ast.reason"
+  else
+    echo "PASS" >"${OUT_DIR}/gate-c.ast.verdict"
+    echo "all new hooks have call-expression in production code" >"${OUT_DIR}/gate-c.ast.reason"
+  fi
+}
+
+# ---------------------------------------------------------------------------
 # Gate D — Mock-coverage delta (advisory)
 # ---------------------------------------------------------------------------
 

--- a/scripts/orch-reviewer-run.sh
+++ b/scripts/orch-reviewer-run.sh
@@ -287,7 +287,6 @@ gate_c() {
 # in any executable position; type annotations, type-only imports, and
 # comments do not match. Files under `__tests__/` and any path ending in
 # `.test.ts` are filtered out so test-only call sites do not count.
-# shellcheck disable=SC2329  # invoked from gate_c_ast; dispatched in a follow-up plan item
 has_production_call_expr() {
   local name="$1"
   local matches
@@ -297,7 +296,6 @@ has_production_call_expr() {
   [[ -n "${matches}" ]]
 }
 
-# shellcheck disable=SC2329  # added to dispatch in a follow-up plan item
 gate_c_ast() {
   local hooks unwired=""
   hooks=$(exported_hooks || true)
@@ -380,11 +378,13 @@ gate_d() {
 gate_a
 gate_b
 gate_c
+gate_c_ast
 gate_d
 
 a=$(cat "${OUT_DIR}/gate-a.verdict")
 b=$(cat "${OUT_DIR}/gate-b.verdict")
 c=$(cat "${OUT_DIR}/gate-c.verdict")
+c_ast=$(cat "${OUT_DIR}/gate-c.ast.verdict")
 d=$(cat "${OUT_DIR}/gate-d.verdict")
 
 aggregate="PASS"
@@ -416,6 +416,7 @@ cat >"${VERDICT}" <<JSON
   "gateA": "${a}",
   "gateB": "${b}",
   "gateC": "${c}",
+  "gateCAst": "${c_ast}",
   "gateD": "${d}",
   "aggregate": "${aggregate}",
   "blocking_gates": [${blocking_json}]
@@ -436,6 +437,10 @@ $(cat "${OUT_DIR}/gate-b.reason")
 ## Gate C — Wiring graph
 **Verdict:** ${c}
 $(cat "${OUT_DIR}/gate-c.reason")
+
+## Gate C v2 — Wiring graph (AST-based, advisory)
+**Verdict:** ${c_ast}
+$(cat "${OUT_DIR}/gate-c.ast.reason")
 
 ## Gate D — Mock-coverage delta (advisory)
 **Verdict:** ${d}

--- a/tests/orch/test_gate_c_ast.bats
+++ b/tests/orch/test_gate_c_ast.bats
@@ -1,0 +1,232 @@
+#!/usr/bin/env bats
+#
+# Tests for Gate C v2 — AST-based wiring detector (gate_c_ast).
+#
+# Forensics: PR #262 dogfood (2026-05-07) — the v1 grep-based gate_c
+# false-passed when a newly exported hook was *referenced* in a type
+# position (import, type annotation, re-export) but never *called* from
+# any production code path. v2 augments the gate with an ast-grep check
+# that requires a call-expression `Name(...)` in non-test `.ts` code.
+# Rolled out advisory: the v2 verdict is recorded in
+# `${OUT_DIR}/gate-c.ast.verdict` for observation; the aggregate SHIP/FAIL
+# still gates on the v1 grep verdict (`gate-c.verdict`).
+#
+# Contract under test (added by the plan to scripts/orch-reviewer-run.sh):
+#
+#   gate_c_ast()
+#     - For each name from exported_hooks(), search non-test .ts files
+#       under REPO_ROOT for a call-expression matching the ast-grep
+#       pattern '$NAME($$$)' under --lang ts.
+#     - PASS  → every hook has ≥1 call-expression match in production code
+#     - FAIL  → ≥1 hook has no call-expression match (type-only references
+#               and pure absence both fall here)
+#     - SKIP  → ast-grep is not on PATH (fail-open, advisory)
+#     - PASS  → no exported hooks introduced (nothing to check)
+#     - Writes the verdict to ${OUT_DIR}/gate-c.ast.verdict and a one-line
+#       reason to ${OUT_DIR}/gate-c.ast.reason.
+#     - Does NOT modify gate-c.verdict (the v1 aggregate input).
+
+setup() {
+  TEST_TMP="$(mktemp -d -t orch-gate-c-ast-XXXXXX)"
+  export TEST_TMP
+  REPO_ROOT_REAL="$(cd "${BATS_TEST_DIRNAME}/../.." && pwd)"
+  RUNNER="${REPO_ROOT_REAL}/scripts/orch-reviewer-run.sh"
+  mkdir -p "${TEST_TMP}/repo" "${TEST_TMP}/out"
+}
+
+teardown() {
+  if [[ -n "${TEST_TMP:-}" && -d "${TEST_TMP}" ]]; then
+    rm -rf "${TEST_TMP}"
+  fi
+}
+
+# A minimal plan body: no decision-log rows, so gate_b passes trivially
+# and the aggregate verdict is determined by gate_a/c/d only.
+write_plan() {
+  cat >"${TEST_TMP}/plan.md" <<'EOF'
+# Plan
+
+## Approach
+
+Adds a hook for downstream wiring.
+EOF
+}
+
+# Synthesize a unified diff that adds an exported hook in runner.ts.
+# The hook name must match the regex in exported_hooks() — the suffix
+# (Hook|Callback|Handler|Adapter|Client) or `On<UpperCase>...` prefix.
+write_diff_with_hook() {
+  local hook_name="$1"
+  cat >"${TEST_TMP}/diff.patch" <<EOF
+diff --git a/runner.ts b/runner.ts
+--- a/runner.ts
++++ b/runner.ts
+@@ -10,6 +10,7 @@
++export type ${hook_name} = (input: { value: number }) => void;
+EOF
+}
+
+# Run the reviewer with the test fixtures in place. Wraps `set +e` so a
+# FAIL aggregate (exit 1) does not abort the test — the AST verdict is
+# advisory and we only inspect gate-c.ast.verdict.
+run_reviewer() {
+  set +e
+  bash "${RUNNER}" \
+    --plan "${TEST_TMP}/plan.md" \
+    --diff "${TEST_TMP}/diff.patch" \
+    --repo-root "${TEST_TMP}/repo" \
+    --out "${TEST_TMP}/out" \
+    >/dev/null 2>&1
+  set -e
+}
+
+# Build a PATH that excludes every directory containing an `ast-grep`
+# binary. Used by the SKIP test to simulate ast-grep not being installed
+# without actually uninstalling it.
+path_without_ast_grep() {
+  local sanitized=""
+  local IFS=':'
+  for d in ${PATH}; do
+    [[ -z "${d}" ]] && continue
+    if [[ ! -x "${d}/ast-grep" ]]; then
+      sanitized="${sanitized}${sanitized:+:}${d}"
+    fi
+  done
+  printf '%s' "${sanitized}"
+}
+
+# --- ast-grep available: verdicts driven by detection rule ---
+
+@test "type-only reference produces FAIL on AST verdict" {
+  command -v ast-grep >/dev/null 2>&1 || skip "ast-grep not installed"
+  write_plan
+  write_diff_with_hook "OnSampleHook"
+  # Production caller mentions the hook only as a type — never calls it.
+  cat >"${TEST_TMP}/repo/wiring.ts" <<'TS'
+import type { OnSampleHook } from "./runner.js";
+
+const handler: OnSampleHook = (input) => {
+  console.log(input.value);
+};
+export { handler };
+TS
+  run_reviewer
+  [ -f "${TEST_TMP}/out/gate-c.ast.verdict" ]
+  [ "$(cat "${TEST_TMP}/out/gate-c.ast.verdict")" = "FAIL" ]
+}
+
+@test "call-expression in non-test code produces PASS on AST verdict" {
+  command -v ast-grep >/dev/null 2>&1 || skip "ast-grep not installed"
+  write_plan
+  write_diff_with_hook "InvokeFooClient"
+  # Production caller invokes the symbol as a call expression.
+  cat >"${TEST_TMP}/repo/wiring.ts" <<'TS'
+import { InvokeFooClient } from "./runner.js";
+
+InvokeFooClient({ value: 42 });
+TS
+  run_reviewer
+  [ -f "${TEST_TMP}/out/gate-c.ast.verdict" ]
+  [ "$(cat "${TEST_TMP}/out/gate-c.ast.verdict")" = "PASS" ]
+}
+
+@test "call-expression only in __tests__ does not count → FAIL on AST verdict" {
+  command -v ast-grep >/dev/null 2>&1 || skip "ast-grep not installed"
+  write_plan
+  write_diff_with_hook "OrphanHandler"
+  # The only call site is under __tests__/ — must be ignored by the AST
+  # detector, leaving zero production callers.
+  mkdir -p "${TEST_TMP}/repo/__tests__"
+  cat >"${TEST_TMP}/repo/__tests__/runner.test.ts" <<'TS'
+import { OrphanHandler } from "../runner.js";
+
+OrphanHandler({ value: 1 });
+TS
+  run_reviewer
+  [ -f "${TEST_TMP}/out/gate-c.ast.verdict" ]
+  [ "$(cat "${TEST_TMP}/out/gate-c.ast.verdict")" = "FAIL" ]
+}
+
+@test "no references anywhere produces FAIL on AST verdict" {
+  command -v ast-grep >/dev/null 2>&1 || skip "ast-grep not installed"
+  write_plan
+  write_diff_with_hook "OnUnreferencedHook"
+  # Repo has a non-test .ts file that does not mention the hook at all.
+  cat >"${TEST_TMP}/repo/unrelated.ts" <<'TS'
+export const noop = () => {};
+TS
+  run_reviewer
+  [ -f "${TEST_TMP}/out/gate-c.ast.verdict" ]
+  [ "$(cat "${TEST_TMP}/out/gate-c.ast.verdict")" = "FAIL" ]
+}
+
+# --- ast-grep missing: SKIP (fail-open, advisory) ---
+
+@test "ast-grep missing from PATH produces SKIP on AST verdict" {
+  write_plan
+  write_diff_with_hook "OnSkipHook"
+  # Even with a "good" production caller, a missing ast-grep binary
+  # forces SKIP — the gate cannot answer, so it must not emit FAIL.
+  cat >"${TEST_TMP}/repo/wiring.ts" <<'TS'
+import { OnSkipHook } from "./runner.js";
+
+OnSkipHook({ value: 1 });
+TS
+  local sanitized
+  sanitized="$(path_without_ast_grep)"
+  set +e
+  PATH="${sanitized}" bash "${RUNNER}" \
+    --plan "${TEST_TMP}/plan.md" \
+    --diff "${TEST_TMP}/diff.patch" \
+    --repo-root "${TEST_TMP}/repo" \
+    --out "${TEST_TMP}/out" \
+    >/dev/null 2>&1
+  set -e
+  [ -f "${TEST_TMP}/out/gate-c.ast.verdict" ]
+  [ "$(cat "${TEST_TMP}/out/gate-c.ast.verdict")" = "SKIP" ]
+}
+
+@test "ast-grep missing → reason mentions ast-grep so logs are diagnosable" {
+  write_plan
+  write_diff_with_hook "OnSkipHook"
+  cat >"${TEST_TMP}/repo/wiring.ts" <<'TS'
+import { OnSkipHook } from "./runner.js";
+OnSkipHook({ value: 1 });
+TS
+  local sanitized
+  sanitized="$(path_without_ast_grep)"
+  set +e
+  PATH="${sanitized}" bash "${RUNNER}" \
+    --plan "${TEST_TMP}/plan.md" \
+    --diff "${TEST_TMP}/diff.patch" \
+    --repo-root "${TEST_TMP}/repo" \
+    --out "${TEST_TMP}/out" \
+    >/dev/null 2>&1
+  set -e
+  [ -f "${TEST_TMP}/out/gate-c.ast.reason" ]
+  grep -q 'ast-grep' "${TEST_TMP}/out/gate-c.ast.reason"
+}
+
+# --- aggregate isolation: AST verdict is advisory only ---
+
+@test "AST FAIL does not flip aggregate when v1 grep gate_c PASSes" {
+  command -v ast-grep >/dev/null 2>&1 || skip "ast-grep not installed"
+  write_plan
+  write_diff_with_hook "OnAdvisoryHook"
+  # Type-only reference: v1 grep finds the symbol → gate-c.verdict=PASS.
+  # v2 ast-grep finds no call expression → gate-c.ast.verdict=FAIL.
+  # Aggregate must remain PASS because Gate C v2 is advisory.
+  cat >"${TEST_TMP}/repo/wiring.ts" <<'TS'
+import type { OnAdvisoryHook } from "./runner.js";
+
+const handler: OnAdvisoryHook = (input) => {
+  console.log(input.value);
+};
+export { handler };
+TS
+  run_reviewer
+  [ "$(cat "${TEST_TMP}/out/gate-c.verdict")" = "PASS" ]
+  [ "$(cat "${TEST_TMP}/out/gate-c.ast.verdict")" = "FAIL" ]
+  # verdict.json's aggregate must not be FAIL just because AST FAILed.
+  grep -q '"aggregate": "PASS"' "${TEST_TMP}/out/verdict.json"
+}

